src/main/java/org/opensrp/web/rest/ClientFormResource.java

```diff
@@ -230,17 +231,127 @@ private ClientFormMetadata getMostRecentVersion(@NonNull String formIdentifier,
+                    HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        return new ResponseEntity<>(HttpStatus.CREATED);
+    }
+
+    private String getFormVersion(String formVersion){
+        String version = formVersion;
+        if (StringUtils.isBlank(formVersion)) {
+            version = generateFormVersion(2);
+        }
+        return  version;
+    }
+
+    private String getIdentifier(String formIdentifier, MultipartFile jsonFile) {
+        String identifier = formIdentifier;
+        if (StringUtils.isBlank(formIdentifier)) {
+            identifier = Paths.get(jsonFile.getOriginalFilename()).getFileName().toString();
+        }
+        return identifier;
+    }
+
+    private ClientForm getClientForm(String fileContentString) {
+        ClientForm clientForm = new ClientForm();
+        clientForm.setJson(fileContentString);
+        clientForm.setCreatedAt(new Date());
+        return clientForm;
+    }
+
+    private String generateFormVersion(int limit) {
+        String formVersion = "0.0.1";
+        if (limit < 6) {//I don't think we would have more than six manifests without form versions.
+            List<Manifest> manifestList = manifestService.getAllManifest(limit);
+            if (manifestList != null && manifestList.size() > 0) {
+                for (int i = 0; i < manifestList.size(); i++) {
+                    Manifest manifest = manifestList.get(i);
+                    String json = manifest.getJson();
+                    if (StringUtils.isNotBlank(json)) {
+                        JSONObject jsonObject = new JSONObject(json);
+                        if (jsonObject.has(FORMS_VERSION)) {
+                            String version = jsonObject.getString(FORMS_VERSION);
+                            if (StringUtils.isNotBlank(version)) {
+                                DefaultArtifactVersion defaultArtifactVersion = new DefaultArtifactVersion(version);
+                                if (defaultArtifactVersion.getIncrementalVersion() < 1000) {
+                                    int newVersion = defaultArtifactVersion.getIncrementalVersion() + 1;
+                                    formVersion =
+                                            defaultArtifactVersion.getMajorVersion() + "." + defaultArtifactVersion.getMinorVersion() +
+                                                    "." + newVersion;
+                                } else if (defaultArtifactVersion.getMinorVersion() < 1000) {
+                                    int newVersion = defaultArtifactVersion.getMinorVersion() + 1;
+                                    formVersion =
+                                            defaultArtifactVersion.getMajorVersion() + "." + newVersion + "." + defaultArtifactVersion.getIncrementalVersion();
+                                } else {
+                                    int newVersion = defaultArtifactVersion.getMajorVersion() + 1;
```

Comment:
Create an issue to create a bug that fixes a situation where `2.1000.1000` & the next ends up being `3.1000.1000`. I think this should start back at  `2.0.1` or rather where `2.0.1000` and next should be `2.1.0`

This issue will also include any other cleanup for this PR

Reply:
makes sense.

Reply:
https://github.com/OpenSRP/opensrp-server-web/issues/398